x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,440 @@
+	}
+
+	k.DeleteDenylist(ctx, consumerId)
+	for _, address := range updateRecord.Denylist {
+		consAddr, err := sdk.ConsAddressFromBech32(address)
+		if err != nil {
+			continue
+		}
+
+		k.SetDenylist(ctx, consumerId, types.NewProviderConsAddress(consAddr))
+	}
+
+	oldTopN, found := k.GetTopN(ctx, consumerId)
+	if !found {
+		oldTopN = 0
+		k.Logger(ctx).Info("consumer chain top N not found, treating as 0", "consumerId", consumerId)
+	}
+
+	// if the top N changes, we need to update the new minimum power in top N
+	if updateRecord.Top_N != oldTopN {
+		if updateRecord.Top_N > 0 {
+			// if the chain receives a non-zero top N value, store the minimum power in the top N
+			bondedValidators, err := k.GetLastBondedValidators(ctx)
+			if err != nil {
+				return err
+			}
+			minPower, err := k.ComputeMinPowerInTopN(ctx, bondedValidators, updateRecord.Top_N)
+			if err != nil {
+				return err
+			}
+			k.SetMinimumPowerInTopN(ctx, consumerId, minPower)
+		} else {
+			// if the chain receives a zero top N value, we delete the min power
+			k.DeleteMinimumPowerInTopN(ctx, consumerId)
+		}
+	}
+
+	k.SetMinStake(ctx, consumerId, updateRecord.MinStake)
+	k.SetInactiveValidatorsAllowed(ctx, consumerId, updateRecord.AllowInactiveVals)
+
+	return nil
+}
+
+// GetLaunchedConsumersReadyToStop returns the consumer ids of the pending launched consumer chains
+// that are ready to stop
+func (k Keeper) GetLaunchedConsumersReadyToStop(ctx sdk.Context) []string {
+	store := ctx.KVStore(k.storeKey)
+	iterator := storetypes.KVStorePrefixIterator(store, types.ConsumerIdToStopTimeKeyNamePrefix())
+	defer iterator.Close()
+
+	var consumerIds []string
+
+	for ; iterator.Valid(); iterator.Next() {
+		// the `consumerId` resides in the whole key, but we skip the first byte (because it's the `ConsumerIdKey`)
+		// plus 8 more bytes for the `uint64` in the key that contains the length of the `consumerId`
+		consumerId := string(iterator.Key()[1+8:])
+
+		var stopTime time.Time
+		err := stopTime.UnmarshalBinary(iterator.Value())
+		if err != nil {
+			panic(fmt.Errorf("failed to unmarshal stop stopTime: %w for consumer id: %s", err, consumerId))
+		}
+
+		// if current block time is equal to or after stop stopTime, and the chain is launched we can stop the chain
+		phase, found := k.GetConsumerIdToPhase(ctx, consumerId)
+		if !ctx.BlockTime().Before(stopTime) && (found && phase == Launched) {
+			consumerIds = append(consumerIds, string(iterator.Key()[1+8:]))
+		}
+	}
+
+	return consumerIds
+}
+
+// IsValidatorOptedInToChain checks if the validator with `providerAddr` is opted into the chain with the specified `chainId`.
+// It returns `found == true` and the corresponding chain's `consumerId` if the validator is opted in. Otherwise, it returns an empty string
+// for `consumerId` and `found == false`.
+func (k Keeper) IsValidatorOptedInToChain(ctx sdk.Context, providerAddr types.ProviderConsAddress, chainId string) (consumerId string, found bool) {
```

Comment:
```suggestion
func (k Keeper) IsValidatorOptedInToChainId(ctx sdk.Context, providerAddr types.ProviderConsAddress, chainId string) (consumerId string, found bool) {
```

Reply:
> We should have an index from `chainId` to a list of `consumerIds`. Ideally, the list will contain only one element, but given the permissionless nature this will not be necessarily true.

An index from `chainId` to `consumerId`s might become extremely large (e.g., with 1M registrations for a specific chain). So I created an index from a validator `providerAddr` to its opted-in `consumerId` chains instead. Such an index can still become arbitrarily large but this implies the validator has opted-in to a multitude of chains.
If you're okay with this, feel free to resolve this conversation.

Reply:
That's the exact case we want to prevent -- malicious behavior that would try to scam validators into believing they are opting in on a real consumer chain.